# Ford Go Bike **2019** Dataset Wrangling and visualisation 
> Authored by: **Abhishek Pandey** [LinkedIn](https://www.linkedin.com/in/abhishekpandeyit/) | [Twitter](https://twitter.com/PandeyJii_)

## Introduction
> This data set includes information about individual rides made in a bike-sharing system covering the greater San Francisco Bay area.
>
> [**Bay Wheels**](https://en.wikipedia.org/wiki/Bay_Wheels#:~:text=Bay%20Wheels%20is%20a%20regional,Area%20Air%20Quality%20Management%20District.) is a regional public bicycle sharing system in the San Francisco Bay Area, California operated by Motivate in a partnership with the Metropolitan Transportation Commission and the Bay Area Air Quality Management District. Bay Wheels is the first regional and large-scale bicycle sharing system deployed in California and on the West Coast of the United States. It was established as Bay Area Bike Share in August 2013. As of January 2018, the Bay Wheels system had over 2,600 bicycles in 262 stations across San Francisco, East Bay and San Jose.

In [2]:
"""
Importing all the required python libraries for wrangling process
"""
import numpy as np #For numerical 
import pandas as pd
import matplotlib.pyplot as plt #For plotting graphs
import seaborn as sb
import os
import glob
%matplotlib inline

## Data Gathering

In [49]:
"""Appending all datasets together to intiate the process of wrangling"""

folder_name = 'Datasets'
path_to_save = "F:/Github_repo/BayWheelsRideDataExploration/final_dataset_2019.csv"

def append():
    frames = [pd.read_csv(D) for D in glob.glob(os.path.join(folder_name, '*.csv'))]
    data_concat = pd.concat(frames, ignore_index=True) #Concatenating all the csv files using pandas concat function.
    data_concat.to_csv(path_to_save, index=False) #Saving the concatenated dataset to a single file for further exploraton.
    data_concat.shape

append()    

In [4]:
df = pd.read_csv('F:/Github_repo/BayWheelsRideDataExploration/final_dataset_2019.csv')
df.head()

duration_sec                start_time                  end_time  \
0         80825  2019-01-31 17:57:44.6130  2019-02-01 16:24:49.8640   
1         65900  2019-01-31 20:58:33.8860  2019-02-01 15:16:54.1730   
2         62633  2019-01-31 18:06:52.9240  2019-02-01 11:30:46.5300   
3         44680  2019-01-31 19:46:09.7190  2019-02-01 08:10:50.3180   
4         60709  2019-01-31 14:19:01.5410  2019-02-01 07:10:51.0650   

   start_station_id           start_station_name  start_station_latitude  \
0             229.0    Foothill Blvd at 42nd Ave               37.775745   
1               4.0  Cyril Magnin St at Ellis St               37.785881   
2             245.0       Downtown Berkeley BART               37.870139   
3              85.0      Church St at Duboce Ave               37.770083   
4              16.0      Steuart St at Market St               37.794130   

   start_station_longitude  end_station_id              end_station_name  \
0              -122.213037           196.0       Grand Ave at Perkins St   
1              -122.408915           134.0        Valencia St at 24th St   
2              -122.268422           157.0          65th St at Hollis St   
3              -122.429156            53.0        Grove St at Divisadero   
4              -122.394430            28.0  The Embarcadero at Bryant St   

   end_station_latitude  end_station_longitude  bike_id   user_type  \
0             37.808894            -122.256460     4861  Subscriber   
1             37.752428            -122.420628     5506  Subscriber   
2             37.846784            -122.291376     2717    Customer   
3             37.775946            -122.437777     4557    Customer   
4             37.787168            -122.388098     2100    Customer   

  bike_share_for_all_trip rental_access_method  
0                      No                  NaN  
1                      No                  NaN  
2                      No                  NaN  
3                      No                  NaN  
4                      No                  NaN

In [41]:
'''Defining function to do some common exploration'''
dataframe = df
def explore(mode):
    if mode == "size":
        print(f"Rows ={dataframe.shape[0]} and Columns={dataframe.shape[1]} ")
    if mode == "info":
        print("Column-Datatype Information\n")
        dataframe.info()
    if mode == "desc":
        print("Common Description\n")
        dataframe.describe()
    if mode == "null":
        print("Columns with Null/NAN Value\n")
        return dataframe.isnull().sum()
    if mode== "dup":
        print(f"No of Duplicate values = {dataframe.duplicated().sum()}")
    

In [6]:
explore("size")

Rows =2506983 and Columns=15 


In [7]:
explore('info')

Column-Datatype Information

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2506983 entries, 0 to 2506982
Data columns (total 15 columns):
 #   Column                   Dtype  
---  ------                   -----  
 0   duration_sec             int64  
 1   start_time               object 
 2   end_time                 object 
 3   start_station_id         float64
 4   start_station_name       object 
 5   start_station_latitude   float64
 6   start_station_longitude  float64
 7   end_station_id           float64
 8   end_station_name         object 
 9   end_station_latitude     float64
 10  end_station_longitude    float64
 11  bike_id                  int64  
 12  user_type                object 
 13  bike_share_for_all_trip  object 
 14  rental_access_method     object 
dtypes: float64(6), int64(2), object(7)
memory usage: 286.9+ MB


In [8]:
explore("null")

Columns with Null/NAN Value



duration_sec                     0
start_time                       0
end_time                         0
start_station_id             80734
start_station_name           80133
start_station_latitude           0
start_station_longitude          0
end_station_id               82902
end_station_name             82350
end_station_latitude             0
end_station_longitude            0
bike_id                          0
user_type                        0
bike_share_for_all_trip     243259
rental_access_method       2386145
dtype: int64

In [42]:
explore("dup")

No of Duplicate values = 0


### Finding Value counts of the possible columns in our Data Set

In [48]:
columns=["user_type", "bike_share_for_all_trip","bike_id"]
for col in columns:
    print(f"Value counts for {col}")
    print(df[col].value_counts())
    print("\n\n")

Value counts for user_type
Subscriber    2021166
Customer       485817
Name: user_type, dtype: int64



Value counts for bike_share_for_all_trip
No     2086152
Yes     177572
Name: bike_share_for_all_trip, dtype: int64





**Finally our process of Data Gathering completes here and I reach on the following conclusions after exploring our dataset at the initial stage:**
> - The size of this dataset is Rows =2506983 and Columns=15 
> - This dataset do not have any duplicate value but there are many null values present in this dataset.
> - The attributes or columns in this Dataset have incorrect datatypes.

To solve these issue I'll further explore this dataset in the Data Cleaning Process Next to this part.

## Data Cleaning
> Cleaning this dataset might affect our original dataset, so to reduce the dataloss I am creating another copy of our dataframe and will work further on it.

In [51]:
#Creating a copy of our dataframe
df_new=df.copy()

### Fixing the issue of in correct datatypes in our dataframe `df_new`

In [13]:
df_new['start_time'] = pd.to_datetime(trips18['start_time'])
df_new['end_time'] = pd.to_datetime(trips18['end_time'])

df_new['start_station_id'] = trips18['start_station_id'].astype('str')
df_new['end_station_id'] = trips18['end_station_id'].astype('str')
df_new['bike_id'] = trips18['bike_id'].astype('str')

df_new['user_type'] = trips18['user_type'].astype('category')


trips18.info(null_counts=True)

NameError: name 'biketrips18' is not defined

In [14]:
# issue 2: add new columns for trip duration in minute, trip start date in yyyy-mm-dd format, trip start hour of the day, day of week and month

trips18['duration_minute'] = trips18['duration_sec']/60

trips18['start_date'] = trips18.start_time.dt.strftime('%Y-%m-%d')
trips18['start_hourofday'] = trips18.start_time.dt.strftime('%H')
trips18['start_dayofweek'] = trips18.start_time.dt.strftime('%A')
trips18['start_month'] = trips18.start_time.dt.strftime('%B')

trips18.head()

NameError: name 'trips18' is not defined

In [15]:
# issue 3: add a new column calculating riders' age from 'member_birth_year'

trips18['member_age'] = 2019 - trips18['member_birth_year']
trips18.describe()

NameError: name 'trips18' is not defined

In [16]:
# plot the distribution of members' age, visually examine for some outliers over 100 yrs old

plt.figure(figsize=[8, 6])
bins = np.arange(0, trips18['member_age'].max()+5, 5)
plt.hist(trips18['member_age'].dropna(), bins=bins);

NameError: name 'trips18' is not defined

In [17]:
# issue 4: filter out outlier ages from visually examination of the distribution above
# issue 5: cast 'member_birth_year' and 'member_age' to integer instead of float type

trips18 = trips18.query('member_age <= 70')
trips18['member_birth_year'] = trips18['member_birth_year'].astype('int')
trips18['member_age'] = trips18['member_age'].astype('int')
trips18.info(null_counts=True)

NameError: name 'trips18' is not defined

### What is the structure of your dataset?

The original combined data contains approximately 1,860,000 individual trip records with 16 variables collected. The variables can be divided into 3 major categories:
- trip duration: `duration_sec`, `start_time`, `end_time`


- station info: `start_station_id`, `start_station_name`, `start_station_latitude`, `start_station_longitude`, `end_station_id`, `end_station_name`, `end_station_latitude`, `end_station_longitude`


- member info (anonymized): `bike_id`, `user_type`, `member_birth_year`, `member_gender`, `bike_share_for_all_trip`

Derived features/variables to assist exploration and analysis:
- trip info: `duration_minute`, `start_date`, `start_hourofday`, `start_dayofweek`, `start_month`


- member: `member_age`

### What is/are the main feature(s) of interest in your dataset?

I'm most interested in exploring the bike trips' duration and rental events occurrance patterns, along with how these relate to the riders' characteristics, i.e. their user type, gender, age, etc, to get a sense of how and what people are using the bike sharing service for. Sample questions to answer: When are most trips taken in terms of time of day, day of the week, or month of the year? How long does the average trip take? Does the above depend on if a user is a subscriber or customer?

### What features in the dataset do you think will help support your investigation into your feature(s) of interest?

Each trip's start date/time and duration information will help understanding how long a trip usually takes and when. The member information like user type, gender and age will help us find out who are the main target customer groups, use the different groups to summarize bike usage data to see if there is any special pattern associated with a specific group of riders.

## Univariate Exploration

A series of plots to first explore the trips distribution over hour-of-day, day-of-week and month.

In [18]:
# trip distribution over day hours

plt.rcParams['figure.figsize'] = 8, 6
base_color = sb.color_palette('colorblind')[0]
sb.set_style('darkgrid')

sb.countplot(data=trips18, x='start_hourofday', color=base_color);
plt.xlabel('Trip Start Hour of Day');
plt.ylabel('Count');

NameError: name 'trips18' is not defined

In [19]:
# trip distribution over weekdays
# issue 6: cast 'start_dayofweek' to category dtype

weekday = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
weekdaycat = pd.api.types.CategoricalDtype(ordered=True, categories=weekday)
trips18['start_dayofweek'] = trips18['start_dayofweek'].astype(weekdaycat)

sb.countplot(data=trips18, x='start_dayofweek', color=base_color);
plt.xlabel('Trip Start Day of Week');
plt.ylabel('Count');

NameError: name 'trips18' is not defined

In [20]:
# trip distribution over months
# issue 7: cast 'start_month' to category dtype for easy plotting

month = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December']
monthcat = pd.api.types.CategoricalDtype(ordered=True, categories=month)
trips18['start_month'] = trips18['start_month'].astype(monthcat)

sb.countplot(data=trips18, x='start_month', color=base_color);
plt.xticks(rotation=30);
plt.xlabel('Trip Start Month');
plt.ylabel('Count');

NameError: name 'trips18' is not defined

The trip distribution over day hours peaks around two timeframes, 8am-9am and 17pm-18pm, during typical rush hours. Looking in combined with the trip distribution over day of week plot, it is quite obvious that the majority of rides happened on work days (Mon-Fri) and primary usage is probably for commuting. In the 12 months in 2018, October had the most ride trips compared to the others, but overall it was the most popular during summer time (May-Sept), probably due to the weather in the area.   

The next several plots are around members/users to see what makes up of the riders.

In [21]:
sb.countplot(data=trips18, x='user_type', color=base_color);
plt.xlabel('User Type');
plt.ylabel('Count');

NameError: name 'trips18' is not defined

In [22]:
sb.countplot(data=trips18, x='member_gender', color=base_color);
plt.xlabel('Gender');
plt.ylabel('Count');

NameError: name 'trips18' is not defined

In [23]:
sb.countplot(data=trips18, x='bike_share_for_all_trip', color=base_color);
plt.xlabel('Bike Share for All Trip');
plt.ylabel('Count');

NameError: name 'trips18' is not defined

In [24]:
bins = np.arange(0, trips18['member_age'].max()+5, 5)
plt.hist(data=trips18, x='member_age', bins=bins);
plt.xticks(bins, bins);
plt.xlabel('Member Age');
plt.ylabel('Count');

NameError: name 'trips18' is not defined

Most riders were male subscribers who did not use bike share for all trips. Most members were around 25 to 40 years old, corresponding to the trip distribution plots that most trips were on Mon-Fri and maxed during rush hours of a day. As the age gets older, bike usage dropped significantly. 

Trip duration distribution to plot next. 

In [25]:
plt.hist(data=trips18, x='duration_minute');
plt.xlabel('Trip Duration in Minute');

NameError: name 'trips18' is not defined

It was a little surprising to see the duration distribution spans pretty far to the right (indicates a long time rental, almost 24 hrs) while the majority of the rides were quite short (within 200 mins). Going to zoom in to the most cases for a closer picture of what the distribution is like.   

In [26]:
trips18['duration_minute'].describe(percentiles=[.99])

NameError: name 'trips18' is not defined

In [27]:
bins = np.arange(0, 66, 1)
ticks = np.arange(0, 66, 5)
plt.hist(data=trips18, x='duration_minute', bins=bins);
plt.xticks(ticks, ticks);
plt.xlabel('Trip Duration in Minute');

NameError: name 'trips18' is not defined

It looks like 99% of the trips were less than 1 hour long with most of them fell between 5 to 10 minutes range. I'll remove the outlier long trips for this study to allow focusing on the most common cases to explore the pattern.

In [28]:
# issue 8: filter out outlier trip records where the duration was very long

trips18 = trips18.query('duration_minute <= 66')
trips18.info(null_counts=True)

NameError: name 'trips18' is not defined

### Discuss the distribution(s) of your variable(s) of interest. Were there any unusual points? Did you need to perform any transformations?

The number of trips peaked around 8-9am and 17-18pm during a day, there were more trips on work days (Mon-Fri) compared to weekends. Summar time was the most popular season of a year, likely due to the weather. 

User-wise, there were more male riders than female, and most members were subscribers compared to casual riders. The majority of the members did not use bike share for all of their trips, and most were around 25 to 40 years old. 

Most rides were quick and short, lasted between 5 to 10 minutes, though there were some very long outliers like 24hrs.

No transformation was needed luckily due to the straightforwardness of the data.

### Of the features you investigated, were there any unusual distributions? Did you perform any operations on the data to tidy, adjust, or change the form of the data? If so, why did you do this?

The distribution of the trip duration was a little surprising at first plotting out the original data. Used .describe() and xticks() to zoom in to a smaller range where most trip records fell, which made the distribution much clearer. In order to help focusing on the main cases of typical use of the system, I filtered out the outlier long trip records for the study with .query(). 

In [29]:
# save the clean data to a .csv file

trips18.to_csv('fordgobike_trips_2018_clean.csv', index=False)

NameError: name 'trips18' is not defined

## Bivariate Exploration

How does the trip duration distribution vary between customers and subscribers?

In [30]:
sb.violinplot(data=trips18, x='user_type', y='duration_minute', color=base_color, inner='quartile');
plt.xlabel('User Type');
plt.ylabel('Trip Duration in Minute');

NameError: name 'trips18' is not defined

The trip duration distribution is much narrower for subscribers compared to casual riders on the shorter/quicker trip end overall. It seems like subscribers have a more specific usage or targeted goal riding the bikes compared to customers who vary more and generally rented the bikes for longer.   

How does the trip duration distribution vary by gender?

In [31]:
sb.boxplot(data=trips18, x='member_gender', y='duration_minute', color=base_color);
plt.xlabel('Gender');
plt.ylabel('Trip Duration in Minute');

NameError: name 'trips18' is not defined

Though not a huge difference, male riders tend to have shorter trips compared to female users, indicated by both a smaller median and shorter IQR.  

Average Trip Duration on Weekdays

In [32]:
sb.barplot(data=trips18, x='start_dayofweek', y='duration_minute', color=base_color);
plt.xlabel('Day of Week');
plt.ylabel('Avg. Trip Duration in Minute');

NameError: name 'trips18' is not defined

The riding trips are much shorter on Monday through Friday compared to weekends. It indicates a pretty stable and efficient usage of the sharing system on normal work days, while more casual flexible use on weekends. 

Average trip duration by month

In [33]:
sb.barplot(data=trips18, x='start_month', y='duration_minute', color=base_color);
plt.xticks(rotation=30);
plt.xlabel('Month');
plt.ylabel('Avg. Trip Duration in Minute');

NameError: name 'trips18' is not defined

Starting from March, the average usage time shown an increasing trend almost all the way til October before it down turned. The average longest trips happened during the summer months, around June, July and September, which probably has a lot to do with the weather in the area.

Member age by weekdays

In [34]:
sb.boxplot(data=trips18, x='start_dayofweek', y='member_age', color=base_color);
plt.xlabel('Day of Week');
plt.ylabel('Member Age');

NameError: name 'trips18' is not defined

Riders who rented the bikes Monday through Friday are slightly older than those who ride on weekends, which corresponds to the commute to work usage that was observed from the above univariable exploration plots.  

Weekly usage between customers and subscribers

In [35]:
sb.countplot(data=trips18, x='start_dayofweek', hue='user_type');
plt.xlabel('Day of Week');
plt.ylabel('Count');

NameError: name 'trips18' is not defined

There was much more subscriber usage than casual customers overall. The drop of volume on weekends for subscribers indicates that they primarily ride bikes for work commute during work days, whereas almost the opposite pattern of a slight increase of use for customers on weekends demonstrates that the use was more for leisure/touring and relaxing purposes.  

Daily usage between customers and subscribers

In [36]:
sb.countplot(data=trips18, x='start_hourofday', hue='user_type');
plt.xlabel('Hour of Day');
plt.ylabel('Count');

NameError: name 'trips18' is not defined

Subscriber usage clearly peaks out on typical rush hours when people go to work in the morning and getting off work in the afternoon, double confirmed their usage purpose and goal of riding. Similar pattern was not observed among customers who tend to ride most in the afternoon or early evening as for a different purpose than the subscriber riders.

Yearly usage between customers and subscribers

In [37]:
sb.countplot(data=trips18, x='start_month', hue='user_type');
plt.xticks(rotation=30);
plt.xlabel('Month');
plt.ylabel('Count');

NameError: name 'trips18' is not defined

Both subscribers and customers ride the most during the summer months with subscribers maxed out in October and customers peaked in July. The usage was clearly not popular during winter season like November, December and January likely due to the weather.

Member age between customers and subscribers

In [38]:
sb.boxplot(data=trips18, x='user_type', y='member_age', color=base_color);
plt.xlabel('User Type');
plt.ylabel('Member Age');

NameError: name 'trips18' is not defined

Similar to the Member age by weekdays plot, subscribers who ride most often Monday through Friday are slightly older than customers, with a wider range of ages as well. 

### Talk about some of the relationships you observed in this part of the investigation. How did the feature(s) of interest vary with other features in the dataset?

There are a lot more subscriber usage than customers. The riding habit/pattern varies a lot between subscribers and customers. Subscribers use the bike sharing system for work commnute thus most trips were on work days (Mon-Fri) and especially during rush hours (when going to work in the morning and getting off work in the afternoon), whereas customers tend to ride for fun in the afternoon or early evenings over weekends. Subscriber users are slightly older than customer users who tend to take longer rides overall especially during the summer.  

### Did you observe any interesting relationships between the other features (not the main feature(s) of interest)?

It is interesting to see that subscribers are slightly older than customers on average but take much shorter/quicker rides. And October had the most bike riding usage for subscribers throughout the year. 

## Multivariate Exploration

How does the average trip duration vary in weekdays between customers and subscribers?

In [39]:
sb.pointplot(data=trips18, x='start_dayofweek', y='duration_minute', hue='user_type', dodge=0.3, linestyles="");
plt.xlabel('Day of Week');
plt.ylabel('Avg. Trip Duration in Minute');

NameError: name 'trips18' is not defined

It can be seen from the above plot that subscribers ride much shorter/quicker trips compared to customers on each day of the week. Both user types have an obvious increase of trip duration on Saturdays and Sundays over weekends, especially for casual riders. Subscribers usage seems to be more efficient than customers overall and maintained a very consistent average duration Monday through Friday.     

How does the hourly usage vary during weekdays for customers and subscribers?

In [40]:
plt.suptitle('Hourly Usage during Weekdays for Customers and Subscribers')

plt.subplot(2, 1, 1)
customers = trips18.query('user_type == "Customer"')
ct_counts = customers.groupby(['start_dayofweek', 'start_hourofday']).size()
ct_counts = ct_counts.reset_index(name='count')
ct_counts = ct_counts.pivot(index='start_dayofweek', columns='start_hourofday', values='count')
sb.heatmap(ct_counts, cmap='rocket_r');
plt.title('Customer', loc='right');
plt.xlabel('Hour of Day');
plt.ylabel('Day of Week');

plt.subplot(2, 1, 2)
subscribers = trips18.query('user_type == "Subscriber"')
st_counts = subscribers.groupby(['start_dayofweek', 'start_hourofday']).size()
st_counts = st_counts.reset_index(name='count')
st_counts = st_counts.pivot(index='start_dayofweek', columns='start_hourofday', values='count')
sb.heatmap(st_counts, cmap='rocket_r');
plt.title('Subscriber', loc='right');
plt.xlabel('Hour of Day');
plt.ylabel('Day of Week');

NameError: name 'trips18' is not defined

The heatmaps shows clearly very different usage pattern/habit between the two type of riders. Subscribers use the system heavily on work days i.e. Monday through Friday whereas customers ride a lot on weekends, especially in the afternoon. Many trips concentrated around 8-9am and 17-18pm on work days for subscribers when used for work commute, yet customers tend to use more in the late afternoon around 17pm Monday to Friday. Other than usage purpose and habit, it may also have to do with the traffic situation around these rush hours in the area.  

### Talk about some of the relationships you observed in this part of the investigation. Were there features that strengthened each other in terms of looking at your feature(s) of interest?

The multivariate exploration strengthened some of the patterns discovered in the previous bivariate exploration as well as univariate exploration, the relationship between the multiple variables plotted are visualized altogether and information are presented combined. The efficient/short period of usage for subscribers corresponds to their high concentration on rush hours Monday through Friday, indicating the use is primarily for work commute. The more relaxing and flexible pattern of customer use shows that they're taking advantage of the bike sharing system quite differently from the subscribers, heavily over weekends and in the afternoon, for city tour or leisure purpose probably.        

### Were there any interesting or surprising interactions between features?

The interactions between features are all supplementing each other and quite make sense when looked at combined, there's no big surprise observed. The usage habit difference between male and female is not that much or obvious throughout the exploration, which could be related to the imbalanced number of female riders/records compared to male ones. It'd be interesting to see how male and female use the system differently if there were more female data.    